In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
train = pd.read_csv("/content/drive/MyDrive/dsn26/train.csv")
test = pd.read_csv("/content/drive/MyDrive/dsn26/test.csv")

In [3]:
display(train.head())
print("\n\n")
display(test.head())

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format,total_sales
0,row_00000,PRD-PRFP9S,14.252,Low Fat,0.0271,Frozen Foods,81.37,STORE-AGY,45,Large,Tier_3,Standard Supermarket,1764.98
1,row_00001,PRD-PXXK71,7.698,Low Fat,0.0720,HEALTH AND HYGIENE,42.05,STORE-YLW,35,Small,Tier_1,Standard Supermarket,342.13
2,row_00002,PRD-V5MOIJ,14.264,Regular,0.0421,Canned,41.35,STORE-89Z,33,Medium,Tier_1,Standard Supermarket,378.85
3,row_00003,PRD-UN5Z3J,NaN,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket,5595.72
4,row_00004,PRD-6RDQYB,10.338,Regular,0.0120,meat,203.06,STORE-9RG,28,Small,Tier_2,Standard Supermarket,2375.36


,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
0,row_00009,PRD-2WLNC8,8.628,Low Fat,0.0167,household,190.72,STORE-HL7,23,Medium,Tier_3,Superstore
1,row_00015,PRD-LV9PLM,6.993,Low Fat,0.0579,Health and Hygiene,261.07,STORE-9RG,28,Small,Tier_2,Standard Supermarket
2,row_00019,PRD-ET6ZI7,NaN,Low Fat,0.1262,FRUITS AND VEGETABLES,112.50,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
3,row_00020,PRD-6RX35U,14.034,Regular,0.0761,frozen foods,200.71,STORE-DKU,30,NaN,Tier_2,Standard Supermarket
4,row_00023,PRD-I4J38V,13.063,Low Fat,0.0650,Frozen Foods,150.59,STORE-89Z,33,Medium,Tier_1,Standard Supermarket


In [4]:
test_id = test["id"]

y = train["total_sales"]

In [5]:
train.drop([ "total_sales"], axis=1, inplace=True)

In [6]:
n_train = train.shape[0]

In [7]:
combined = pd.concat([train, test])

In [8]:
train.shape[0], test.shape[0], combined.shape[0]

(6818, 1705, 8523)

In [9]:
combined.head()

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
0,row_00000,PRD-PRFP9S,14.252,Low Fat,0.0271,Frozen Foods,81.37,STORE-AGY,45,Large,Tier_3,Standard Supermarket
1,row_00001,PRD-PXXK71,7.698,Low Fat,0.0720,HEALTH AND HYGIENE,42.05,STORE-YLW,35,Small,Tier_1,Standard Supermarket
2,row_00002,PRD-V5MOIJ,14.264,Regular,0.0421,Canned,41.35,STORE-89Z,33,Medium,Tier_1,Standard Supermarket
3,row_00003,PRD-UN5Z3J,NaN,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
4,row_00004,PRD-6RDQYB,10.338,Regular,0.0120,meat,203.06,STORE-9RG,28,Small,Tier_2,Standard Supermarket


In [10]:
combined.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8523 entries, 0 to 1704
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   8523 non-null   object 
 1   product_code         8523 non-null   object 
 2   product_weight_kg    6992 non-null   float64
 3   fat_content          8523 non-null   object 
 4   shelf_visibility     8523 non-null   float64
 5   product_category     8523 non-null   object 
 6   product_price        8523 non-null   float64
 7   store_code           8523 non-null   object 
 8   store_age_years      8523 non-null   int64  
 9   store_size           6113 non-null   object 
 10  store_location_tier  8523 non-null   object 
 11  store_format         8523 non-null   object 
dtypes: float64(3), int64(1), object(8)
memory usage: 865.6+ KB


## Exploratory Data Analysis and Data Cleaning

In [11]:
combined.head()

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
0,row_00000,PRD-PRFP9S,14.252,Low Fat,0.0271,Frozen Foods,81.37,STORE-AGY,45,Large,Tier_3,Standard Supermarket
1,row_00001,PRD-PXXK71,7.698,Low Fat,0.0720,HEALTH AND HYGIENE,42.05,STORE-YLW,35,Small,Tier_1,Standard Supermarket
2,row_00002,PRD-V5MOIJ,14.264,Regular,0.0421,Canned,41.35,STORE-89Z,33,Medium,Tier_1,Standard Supermarket
3,row_00003,PRD-UN5Z3J,NaN,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
4,row_00004,PRD-6RDQYB,10.338,Regular,0.0120,meat,203.06,STORE-9RG,28,Small,Tier_2,Standard Supermarket


In [12]:
combined.isna().sum()

,0
id,0
product_code,0
product_weight_kg,1531
fat_content,0
shelf_visibility,0
product_category,0
product_price,0
store_code,0
store_age_years,0
store_size,2410


### Exploring the `store_size` with missing information

In [13]:
empty_store_size = combined[combined["store_size"].isna()]
empty_store_size.head()

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
9,row_00010,PRD-DH2TA3,8.755,Low Fat,0.0258,household,104.60,STORE-JOR,34,NaN,Tier_3,Corner Shop
11,row_00012,PRD-UX01UP,11.007,Low Fat,0.0304,Snack Foods,234.38,STORE-OYG,25,NaN,Tier_2,Standard Supermarket
14,row_00016,PRD-9USY38,6.997,Low Fat,0.0908,Snack Foods,112.40,STORE-DKU,30,NaN,Tier_2,Standard Supermarket
18,row_00022,PRD-SUDBCI,12.896,Low Fat,0.0420,SNACK FOODS,178.49,STORE-DKU,30,NaN,Tier_2,Standard Supermarket
19,row_00024,PRD-5QVWC9,6.085,Low Fat,0.1235,HOUSEHOLD,261.49,STORE-OYG,25,NaN,Tier_2,Standard Supermarket


In [14]:
# The unique stores with missing store sizes
empty_store_size["store_code"].unique()

array(['STORE-JOR', 'STORE-OYG', 'STORE-DKU'], dtype=object)

In [15]:
# The count of stores with missing store sizes
empty_store_size["store_code"].value_counts()

,count
store_code,
STORE-DKU,929
STORE-OYG,926
STORE-JOR,555


In [16]:
#  The total count of all stores in the dataset
combined["store_code"].value_counts()

,count
store_code,
STORE-7WS,935
STORE-AGY,932
STORE-YLW,930
STORE-89Z,930
STORE-9RG,930
STORE-DKU,929
STORE-HL7,928
STORE-OYG,926
STORE-JOR,555


#### Observations


- `STORE-DKU`: has a total of 929 occurences in the dataset. all occurences have store_size missing
- `STORE-OYG`: has a total of 926 occurences in the dataset. all occurences have store_size missing
- `STORE-JOR`: has a total of 555 occurences in the dataset. all occurences have store_size missing

In [17]:
empty_store_size[empty_store_size["store_code"] == "STORE-DKU"].head()

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
14,row_00016,PRD-9USY38,6.997,Low Fat,0.0908,Snack Foods,112.40,STORE-DKU,30,NaN,Tier_2,Standard Supermarket
18,row_00022,PRD-SUDBCI,12.896,Low Fat,0.0420,SNACK FOODS,178.49,STORE-DKU,30,NaN,Tier_2,Standard Supermarket
27,row_00036,PRD-0QD58X,17.104,Regular,0.0158,Breads,46.03,STORE-DKU,30,NaN,Tier_2,Standard Supermarket
28,row_00038,PRD-99VH0W,7.730,Low Fat,0.0511,Fruits and Vegetables,75.49,STORE-DKU,30,NaN,Tier_2,Standard Supermarket
33,row_00046,PRD-HZXGUJ,8.078,Low Fat,0.0169,DAIRY,52.02,STORE-DKU,30,NaN,Tier_2,Standard Supermarket


In [18]:
empty_store_size[empty_store_size["store_code"] == "STORE-OYG"].head()

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
11,row_00012,PRD-UX01UP,11.007,Low Fat,0.0304,Snack Foods,234.38,STORE-OYG,25,NaN,Tier_2,Standard Supermarket
19,row_00024,PRD-5QVWC9,6.085,Low Fat,0.1235,HOUSEHOLD,261.49,STORE-OYG,25,NaN,Tier_2,Standard Supermarket
21,row_00028,PRD-SB61LG,20.321,Regular,0.0493,Fruits and Vegetables,187.72,STORE-OYG,25,NaN,Tier_2,Standard Supermarket
35,row_00048,PRD-ONGWUL,13.273,Low Fat,0.0406,household,255.41,STORE-OYG,25,NaN,Tier_2,Standard Supermarket
36,row_00050,PRD-8MX1EV,19.180,Low Fat,0.0147,Soft Drinks,51.69,STORE-OYG,25,NaN,Tier_2,Standard Supermarket


In [19]:
empty_store_size[empty_store_size["store_code"] == "STORE-JOR"].head()

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
9,row_00010,PRD-DH2TA3,8.755,Low Fat,0.0258,household,104.60,STORE-JOR,34,NaN,Tier_3,Corner Shop
30,row_00040,PRD-NAMSQE,12.376,Regular,0.2992,BAKING GOODS,79.91,STORE-JOR,34,NaN,Tier_3,Corner Shop
44,row_00061,PRD-4CK9XH,12.451,Low Fat,0.1737,Household,169.29,STORE-JOR,34,NaN,Tier_3,Corner Shop
78,row_00099,PRD-1N8MTZ,19.413,Low Fat,0.0421,Dairy,44.58,STORE-JOR,34,NaN,Tier_3,Corner Shop
96,row_00126,PRD-2EZP7V,12.766,Low Fat,0.0329,Health and Hygiene,101.70,STORE-JOR,34,NaN,Tier_3,Corner Shop


#### Observations


- `STORE-DKU`: is a Standard Supermarket in a tier_2 location
- `STORE-OYG`: is a Standard Supermarket in a tier_2 location
- `STORE-JOR`: is a Corner Shop in a tier_3 location

In [20]:
print("ALL TIER_2 STANDARD SUPERMARKET")
display(combined[(combined["store_location_tier"] == "Tier_2") & (combined["store_format"] == "Standard Supermarket")].head())
print("\n\n")
display(combined[(combined["store_location_tier"] == "Tier_2") & (combined["store_format"] == "Standard Supermarket")]["store_size"].value_counts().reset_index())

ALL TIER_2 STANDARD SUPERMARKET


,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
4,row_00004,PRD-6RDQYB,10.338,Regular,0.0120,meat,203.06,STORE-9RG,28,Small,Tier_2,Standard Supermarket
11,row_00012,PRD-UX01UP,11.007,Low Fat,0.0304,Snack Foods,234.38,STORE-OYG,25,NaN,Tier_2,Standard Supermarket
14,row_00016,PRD-9USY38,6.997,Low Fat,0.0908,Snack Foods,112.40,STORE-DKU,30,NaN,Tier_2,Standard Supermarket
15,row_00017,PRD-GH9N12,11.154,Regular,0.1041,baking goods,103.22,STORE-9RG,28,Small,Tier_2,Standard Supermarket
18,row_00022,PRD-SUDBCI,12.896,Low Fat,0.0420,SNACK FOODS,178.49,STORE-DKU,30,NaN,Tier_2,Standard Supermarket


,store_size,count
0,Small,930


#### Observations

- From the combined dataset, all Standard Supermarket in a Tier_2 location is a small size.
  - This information will be used to fill `store_size` for `STORE-DKU` and `STORE-OYG`

In [21]:
display(combined[combined['store_format'] == "Corner Shop"].head())
print()
print("FROM THE TABLE ABOVE WE CAN SEE THAT A CORNER SHOP CAN BE TIER_1, OR TIER_3")
print("\n\n")
display(combined[combined['store_format'] == "Corner Shop"]["store_size"].value_counts().reset_index())
print("\nFROM THE TABLE ABOVE WE CAN SEE THAT ALL CORNER STORES ARE SMALL")

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
9,row_00010,PRD-DH2TA3,8.755,Low Fat,0.0258,household,104.60,STORE-JOR,34,NaN,Tier_3,Corner Shop
10,row_00011,PRD-ONGWUL,NaN,Low Fat,0.0730,Household,254.12,STORE-T5G,47,Small,Tier_1,Corner Shop
30,row_00040,PRD-NAMSQE,12.376,Regular,0.2992,BAKING GOODS,79.91,STORE-JOR,34,NaN,Tier_3,Corner Shop
39,row_00053,PRD-KFRZWZ,NaN,Low Fat,0.1239,Meat,53.67,STORE-T5G,47,Small,Tier_1,Corner Shop
44,row_00061,PRD-4CK9XH,12.451,Low Fat,0.1737,Household,169.29,STORE-JOR,34,NaN,Tier_3,Corner Shop



FROM THE TABLE ABOVE WE CAN SEE THAT A CORNER SHOP CAN BE TIER_1, OR TIER_3





,store_size,count
0,Small,528



FROM THE TABLE ABOVE WE CAN SEE THAT ALL CORNER STORES ARE SMALL


#### Observations

- From the above, we can see that a corner shop can be located both in a Tier_1 and a Tier_3 location.
- However from the combined dataset, all Corner Shop are small
  - This information will be used to fill `store_size` for STORE-JOR

#### Conclusion

- All three stores with missing `store_size` information will be filled with `Small`

#### Filling missing store_sizes

In [22]:
# Filling missing Store size with small
combined.fillna({"store_size":"Small"}, inplace = True)

### Exploring the `product_weight_kg` with missing information

In [23]:
product_weight_kg_missing = combined[combined["product_weight_kg"].isna()]
product_weight_kg_missing.head(60)

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
3,row_00003,PRD-UN5Z3J,NaN,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
6,row_00006,PRD-MHLD93,NaN,Low Fat,0.1239,Canned,212.23,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
10,row_00011,PRD-ONGWUL,NaN,Low Fat,0.0730,Household,254.12,STORE-T5G,47,Small,Tier_1,Corner Shop
20,row_00025,PRD-8MX1EV,NaN,Low Fat,0.0148,Soft Drinks,51.50,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
39,row_00053,PRD-KFRZWZ,NaN,Low Fat,0.1239,Meat,53.67,STORE-T5G,47,Small,Tier_1,Corner Shop
46,row_00063,PRD-RACGN7,NaN,Low Fat,0.0571,Health and Hygiene,78.40,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
47,row_00064,PRD-IS25HB,NaN,Low Fat,0.0000,Soft Drinks,124.36,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
49,row_00066,PRD-FVHC3D,NaN,Low Fat,0.0729,Seafood,34.86,STORE-9RG,28,Small,Tier_2,Standard Supermarket
54,row_00071,PRD-YUZ87C,NaN,Low Fat,0.0478,Baking Goods,151.04,STORE-T5G,47,Small,Tier_1,Corner Shop
59,row_00076,PRD-UJA86V,NaN,Low Fat,0.1194,Canned,87.70,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket


In [24]:
#  What store sizes usually have missing weights on the product
product_weight_kg_missing["store_size"].value_counts()

,count
store_size,
Medium,957
Small,561
Large,13


In [25]:
# what stores have missing product weight
product_weight_kg_missing["store_code"].value_counts()

,count
store_code,
STORE-7WS,935
STORE-T5G,528
STORE-AGY,13
STORE-89Z,11
STORE-HL7,11
STORE-YLW,10
STORE-9RG,8
STORE-DKU,7
STORE-OYG,5


In [26]:
combined["store_code"].value_counts()

,count
store_code,
STORE-7WS,935
STORE-AGY,932
STORE-YLW,930
STORE-89Z,930
STORE-9RG,930
STORE-DKU,929
STORE-HL7,928
STORE-OYG,926
STORE-JOR,555


#### Observations

- `STORE-7WS`: All products listed has no product_weight_kg

In [27]:
product_weight_kg_missing[product_weight_kg_missing["store_code"] == "STORE-7WS"]

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
3,row_00003,PRD-UN5Z3J,NaN,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
6,row_00006,PRD-MHLD93,NaN,Low Fat,0.1239,Canned,212.23,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
20,row_00025,PRD-8MX1EV,NaN,Low Fat,0.0148,Soft Drinks,51.50,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
46,row_00063,PRD-RACGN7,NaN,Low Fat,0.0571,Health and Hygiene,78.40,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
47,row_00064,PRD-IS25HB,NaN,Low Fat,0.0000,Soft Drinks,124.36,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
...,...,...,...,...,...,...,...,...,...,...,...,...
1687,row_08418,PRD-74AVMI,NaN,Low Fat,0.0877,SNACK FOODS,232.67,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
1689,row_08450,PRD-40V1GK,NaN,Low Fat,0.0259,SNACK FOODS,256.24,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
1697,row_08477,PRD-2VTJXU,NaN,Low Fat,0.1165,Fruits and Vegetables,45.61,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
1698,row_08479,PRD-MU6FGK,NaN,Low Fat,0.0268,Frozen Foods,52.29,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket


In [28]:
combined[combined["product_code"] == "PRD-UN5Z3J"]

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
3,row_00003,PRD-UN5Z3J,NaN,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
3696,row_04647,PRD-UN5Z3J,12.665,Regular,0.0461,SOFT DRINKS,178.14,STORE-9RG,28,Small,Tier_2,Standard Supermarket
5625,row_07062,PRD-UN5Z3J,12.645,Regular,0.0460,Soft Drinks,178.62,STORE-YLW,35,Small,Tier_1,Standard Supermarket
5707,row_07164,PRD-UN5Z3J,13.389,Regular,0.0460,Soft Drinks,175.17,STORE-OYG,25,Small,Tier_2,Standard Supermarket
1036,row_05132,PRD-UN5Z3J,NaN,Regular,0.0000,SOFT DRINKS,172.78,STORE-AGY,45,Large,Tier_3,Standard Supermarket


#### Observations

- The product `PRD-UN5Z3J` has missing `product_weight_kg` values in store `STORE-7WS` but that informatiion exist in other stores like `STORE-9RG`, `STORE-YLW` and `STORE-OYG`

In [29]:
combined[combined["product_code"] == "PRD-MHLD93"]

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
6,row_00006,PRD-MHLD93,NaN,Low Fat,0.1239,Canned,212.23,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
423,row_00544,PRD-MHLD93,6.321,Low Fat,0.1212,CANNED,212.17,STORE-YLW,35,Small,Tier_1,Standard Supermarket
1848,row_02318,PRD-MHLD93,6.131,Low Fat,0.1329,Canned,213.69,STORE-HL7,23,Medium,Tier_3,Superstore
1979,row_02490,PRD-MHLD93,6.336,Low Fat,0.2218,canned,204.07,STORE-JOR,34,Small,Tier_3,Corner Shop
4275,row_05362,PRD-MHLD93,6.364,Low Fat,0.1282,CANNED,202.03,STORE-89Z,33,Medium,Tier_1,Standard Supermarket
4429,row_05551,PRD-MHLD93,6.132,Low Fat,0.1309,Canned,213.12,STORE-9RG,28,Small,Tier_2,Standard Supermarket


In [30]:
combined[combined["product_code"] == "PRD-UN5Z3J"]

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format
3,row_00003,PRD-UN5Z3J,NaN,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket
3696,row_04647,PRD-UN5Z3J,12.665,Regular,0.0461,SOFT DRINKS,178.14,STORE-9RG,28,Small,Tier_2,Standard Supermarket
5625,row_07062,PRD-UN5Z3J,12.645,Regular,0.0460,Soft Drinks,178.62,STORE-YLW,35,Small,Tier_1,Standard Supermarket
5707,row_07164,PRD-UN5Z3J,13.389,Regular,0.0460,Soft Drinks,175.17,STORE-OYG,25,Small,Tier_2,Standard Supermarket
1036,row_05132,PRD-UN5Z3J,NaN,Regular,0.0000,SOFT DRINKS,172.78,STORE-AGY,45,Large,Tier_3,Standard Supermarket


#### Observations

- The product `PRD-MHLD93` has missing `product_weight_kg` values in store `STORE-7WS` but that informatiion exist in other stores like `STORE-9RG`, `STORE-YLW` and `STORE-OYG` etc
- The product `PRD-UN5Z3J` has missing `product_weight_kg` values in store `STORE-7WS` but that informatiion exist in other stores like `STORE-9RG`, `STORE-YLW` and `STORE-OYG` etc

#### Conclusion

- we will fill missing weights of products with the average values of the same product across other stores.

#### Filling `product_weight_kg`

In [31]:
combined.groupby("product_code").agg({"product_weight_kg": "mean"}).reset_index()

,product_code,product_weight_kg
0,PRD-00CVW9,7.620200
1,PRD-00EIUE,8.676250
2,PRD-00V830,9.679000
3,PRD-02NFGI,5.923250
4,PRD-031L36,6.802750
...,...,...
1554,PRD-ZTO1FG,13.936400
1555,PRD-ZVRERW,17.385200
1556,PRD-ZWW54A,20.624833
1557,PRD-ZXY5TE,12.745000


In [32]:
combined["product_weight_kg"] = combined["product_weight_kg"].fillna(

    combined.groupby(["product_code"])["product_weight_kg"].transform("mean")
)

In [33]:
combined.isna().sum()

,0
id,0
product_code,0
product_weight_kg,4
fat_content,0
shelf_visibility,0
product_category,0
product_price,0
store_code,0
store_age_years,0
store_size,0


In [34]:
#  Filling the remianing 4 data points with overall average of the weight
combined["product_weight_kg"].fillna(combined["product_weight_kg"].mean(), inplace = True)

/tmp/ipykernel_2618/2346270861.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  combined["product_weight_kg"].fillna(combined["product_weight_kg"].mean(), inplace = True)


#### Normalizing the `product_category` column
- so all text is in lower case

In [35]:
combined["product_category"] = combined["product_category"].str.lower()

### Feature Engineering

In [36]:
combined['price_vs_cat'] = combined['product_price'] / combined.groupby('product_category')['product_price'].transform('median')
combined['price_rank_in_cat'] = combined.groupby('product_category')['product_price'].rank(pct=True)
combined['price_vs_store'] = combined['product_price'] / combined.groupby('store_code')['product_price'].transform('mean')

combined['price_per_kg'] = combined['product_price'] / combined['product_weight_kg']
combined['weight_vs_cat'] = combined['product_weight_kg'] / combined.groupby('product_category')['product_weight_kg'].transform('median')

combined["vis_vs_product"] = combined['shelf_visibility'] / combined['shelf_visibility'].mean()
combined['vis_x_price'] = combined['shelf_visibility'] * combined['product_price']

In [37]:
combined.head()

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format,price_vs_cat,price_rank_in_cat,price_vs_store,price_per_kg,weight_vs_cat,vis_vs_product,vis_x_price
0,row_00000,PRD-PRFP9S,14.252000,Low Fat,0.0271,frozen foods,81.37,STORE-AGY,45,Large,Tier_3,Standard Supermarket,0.630238,0.234813,0.575641,5.709374,1.115616,0.409863,2.205127
1,row_00001,PRD-PXXK71,7.698000,Low Fat,0.0720,health and hygiene,42.05,STORE-YLW,35,Small,Tier_1,Standard Supermarket,0.326007,0.086538,0.295775,5.462458,0.638414,1.088934,3.027600
2,row_00002,PRD-V5MOIJ,14.264000,Regular,0.0421,canned,41.35,STORE-89Z,33,Medium,Tier_1,Standard Supermarket,0.298018,0.023112,0.294545,2.898906,1.155650,0.636724,1.740835
3,row_00003,PRD-UN5Z3J,12.899667,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket,1.214306,0.732584,1.247467,13.515853,1.091619,0.679071,7.828315
4,row_00004,PRD-6RDQYB,10.338000,Regular,0.0120,meat,203.06,STORE-9RG,28,Small,Tier_2,Standard Supermarket,1.384185,0.868235,1.419477,19.642097,0.838375,0.181489,2.436720


### Encoding data

#### Encoding `store_size`
- Because these stores have an order, we will be encoding with categorical numbers where the larger store takes the larger number
  - Large = 30
  - Medium = 20
  - Small = 10

In [38]:
combined["store_size"] = combined["store_size"].map({"Large": 30, "Medium": 20, "Small": 10})

#### Encoding `store_location_tier`
- Because these store_location_tier have an order, we will be encoding with categorical numbers where the top tier takes the larger number
  - Tier_3 = 30
  - Tier_2 = 20
  - Tier_1 = 10

In [39]:
combined["store_location_tier"] = combined["store_location_tier"].map({"Tier_3": 30, "Tier_2": 20, "Tier_1": 10})

#### Encoding `store_format`
- Because these store_format have an order, we will be encoding with categorical numbers where the bigger store takes the larger number
  - Flagship Hypermarket = 40
  - Superstore = 30
  - Standard Supermarket = 20
  - Corner Shop = 10

In [40]:
combined["store_format"] = combined["store_format"].map({"Flagship Hypermarket":40, "Superstore": 30, "Standard Supermarket": 20, "Corner Shop": 10})

In [41]:
combined["fat_content"] = combined["fat_content"].map({"Low Fat":20, "Regular": 10})


In [42]:
combined.head()

,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format,price_vs_cat,price_rank_in_cat,price_vs_store,price_per_kg,weight_vs_cat,vis_vs_product,vis_x_price
0,row_00000,PRD-PRFP9S,14.252000,20,0.0271,frozen foods,81.37,STORE-AGY,45,30,30,20,0.630238,0.234813,0.575641,5.709374,1.115616,0.409863,2.205127
1,row_00001,PRD-PXXK71,7.698000,20,0.0720,health and hygiene,42.05,STORE-YLW,35,10,10,20,0.326007,0.086538,0.295775,5.462458,0.638414,1.088934,3.027600
2,row_00002,PRD-V5MOIJ,14.264000,10,0.0421,canned,41.35,STORE-89Z,33,20,10,20,0.298018,0.023112,0.294545,2.898906,1.155650,0.636724,1.740835
3,row_00003,PRD-UN5Z3J,12.899667,10,0.0449,soft drinks,174.35,STORE-7WS,47,20,30,40,1.214306,0.732584,1.247467,13.515853,1.091619,0.679071,7.828315
4,row_00004,PRD-6RDQYB,10.338000,10,0.0120,meat,203.06,STORE-9RG,28,10,20,20,1.384185,0.868235,1.419477,19.642097,0.838375,0.181489,2.436720


#### Dropping Columns

In [43]:
combined.drop(["product_code", "product_category", "store_code"], axis =1, inplace = True, errors = "ignore")

In [44]:
combined.head()

,id,product_weight_kg,fat_content,shelf_visibility,product_price,store_age_years,store_size,store_location_tier,store_format,price_vs_cat,price_rank_in_cat,price_vs_store,price_per_kg,weight_vs_cat,vis_vs_product,vis_x_price
0,row_00000,14.252000,20,0.0271,81.37,45,30,30,20,0.630238,0.234813,0.575641,5.709374,1.115616,0.409863,2.205127
1,row_00001,7.698000,20,0.0720,42.05,35,10,10,20,0.326007,0.086538,0.295775,5.462458,0.638414,1.088934,3.027600
2,row_00002,14.264000,10,0.0421,41.35,33,20,10,20,0.298018,0.023112,0.294545,2.898906,1.155650,0.636724,1.740835
3,row_00003,12.899667,10,0.0449,174.35,47,20,30,40,1.214306,0.732584,1.247467,13.515853,1.091619,0.679071,7.828315
4,row_00004,10.338000,10,0.0120,203.06,28,10,20,20,1.384185,0.868235,1.419477,19.642097,0.838375,0.181489,2.436720


In [45]:
train = combined[:n_train]
test = combined[n_train:]

In [46]:
train.to_csv("/content/drive/MyDrive/dsn26/processed_train.csv", index=False)
test.to_csv("/content/drive/MyDrive/dsn26/processed_test.csv", index=False)

In [48]:
y.to_csv("/content/drive/MyDrive/dsn26/target.csv", index=False)